cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+    && _CUDA_VSTD::__type_set_contains_v<_CUDA_VSTD::__make_type_set<_OtherProperties...>, _Properties...>;
+
+  //! @brief Helper to determine what cudaMemcpyKind we need to copy data from another async_mdarray with different
+  template <class... _OtherProperties>
+  static constexpr cudaMemcpyKind __transfer_kind =
+    __select_execution_space<_OtherProperties...> == _ExecutionSpace::__host
+      ? (__is_host_only ? cudaMemcpyHostToHost : cudaMemcpyHostToDevice)
+      : (__is_host_only ? cudaMemcpyDeviceToHost : cudaMemcpyDeviceToDevice);
+
+  //! @brief Helper to return an async_resource_ref to the currently used resource. Used to grow the async_mdarray
+  __resource_ref_t __borrow_resource() const noexcept
+  {
+    return const_cast<__resource_t&>(__buf_.get_memory_resource());
+  }
+
+  //! @brief Copies data from `[__first, __last)` to \p __dest, where \p __first and \p __dest reside in the same memory
+  //! space
+  //! @param __first Pointer to the start of the input segment.
+  //! @param __last Pointer to the end of the input segment.
+  //! @param __dest Pointer to the start of the output segment.
+  //! @returns Pointer equal to `__dest + __count`
+  _CCCL_HIDE_FROM_ABI pointer __copy_same(const_pointer __first, const_pointer __last, pointer __dest)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      return _CUDA_VSTD::copy(__first, __last, __dest);
+    }
+    else
+    {
+      const auto __count = static_cast<size_t>(__last - __first);
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemcpyAsync,
+        "cudax::async_mdarray::__copy_same: failed to copy data",
+        __dest,
+        __first,
+        sizeof(_Tp) * __count,
+        ::cudaMemcpyDeviceToDevice,
+        __buf_.get_stream().get());
+      return __dest + __count;
+    }
+    _CCCL_UNREACHABLE();
+  }
+
+  //! @brief Copies \p __count elements from `[__first, __last)` to \p __dest, where \p __first and \p __dest reside in
+  //! the different memory spaces
+  //! @param __first Pointer to the start of the input segment.
+  //! @param __last Pointer to the end of the input segment.
+  //! @param __dest Pointer to the start of the output segment.
+  //! @param __count The number of elements to be copied.
+  template <class _Iter, cudaMemcpyKind __kind = __is_host_only ? cudaMemcpyHostToHost : cudaMemcpyHostToDevice>
+  _CCCL_HIDE_FROM_ABI void __copy_cross(_Iter __first, _Iter __last, pointer __dest, size_type __count)
+  {
+    _CCCL_IF_CONSTEXPR (__kind == cudaMemcpyHostToHost)
+    {
+      _CUDA_VSTD::copy(__first, __last, __dest);
+    }
+    else _CCCL_IF_CONSTEXPR (!_CUDA_VSTD::contiguous_iterator<_Iter>)
+    { // For non-coniguous iterators we need to copy into temporary host storage to use cudaMemcpy
+      // This should only ever happen when passing in data from host to device
+      _CCCL_ASSERT(__kind == cudaMemcpyHostToDevice, "Invalid use case!");
+      auto __temp = _CUDA_VSTD::get_temporary_buffer<_Tp>(__count).first;
```

Comment:
This will lose any other members of the returned struct; is that OK?

Reply:
Yes, `get_temporary_buffer` returns a `pair<ptr, size>` and we do already know the size